{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "commandline genomic week2",
      "provenance": [],
      "authorship_tag": "ABX9TyNiOZWFddi3yss3/8NpbZwt",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Fellylove/bioinformatics_algorithms_coursera/blob/master/commandline_genomic_week2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#genetic sequences\n",
        "\n",
        " For this project, it is recommended that you use the VMBox virtual environment provided with the Course package and the tools therein. You may also use your own system and software, however make sure that appropriate versions are installed. The answers are compatible with the following versions of the software: samtools v.1.2, bedtools v.2.24.0.\n",
        "\n",
        "As part of a larger project cataloging genetic variation in the plant Arabidopsis thaliana, you sequenced and assembled the genome of one strain (‘wu_0_A’), then mapped back the reads to the assembled genome. The resulting BAM file is included in the package ‘gencommand_proj2_data.tar.gz’. Using SAMtools and BEDtools as well as other Unix commands introduced in this course, examine the files and answer the following questions. NOTE: Input data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.\n",
        "\n",
        "Apply these rules and steps to the questions marked above each rule.\n",
        "\n",
        "Questions 1 - 5:\n",
        "For the original set of alignments (file ‘athal_wu_0_A.bam’)\n",
        "\n",
        "Questions 6 - 10:\n",
        "Extract only the alignments in the range “Chr3:11,777,000-11,794,000”, corresponding to a locus of interest.\n",
        "\n",
        "Questions 11 - 15:\n",
        "Determine general information about the alignment process from the original BAM file.\n",
        "\n",
        "Questions 16 - 20:\n",
        "Using BEDtools, examine how many of the alignments at point 2 overlap exons at the locus of interest. Use the BEDtools ‘-wo’ option to only report non-zero overlaps. The list of exons is given in the included ‘athal_wu_0_A_annot.gtf’ GTF file.\n",
        "\n",
        "1. How many alignments does the set contain?\n",
        "samtools view athal_wu_0_A.bam | wc -l\n",
        "## 221372\n",
        "\n",
        "2. How many alignments show the read’s mate unmapped?\n",
        "samtools view athal_wu_0_A.bam | cut -f7 | grep \"*\" | wc -l\n",
        "## 65521\n",
        "\n",
        "3. How many alignments contain a deletion (D)?\n",
        "samtools view athal_wu_0_A.bam | cut -f6 | grep \"D\" | wc -l\n",
        "## 2451\n",
        "\n",
        "4. How many alignments show the read’s mate mapped to the same chromosome?\n",
        "samtools view athal_wu_0_A.bam | cut -f7 | grep \"=\" | wc -l\n",
        "## 150913\n",
        "\n",
        "5. How many alignments are spliced?\n",
        "samtools view athal_wu_0_A.bam | cut -f6 | grep \"N\" | wc -l\n",
        "## 0\n",
        "\n",
        "6. How many alignments does the set contain?\n",
        "# sort the bam_file\n",
        "samtools sort athal_wu_0_A.bam -o athal_wu_0_A.sort.bam\n",
        "\n",
        "# index the bam_file\n",
        "samtools index athal_wu_0_A.sort.bam\n",
        "\n",
        "# extract range “Chr3:11,777,000-11,794,000”\n",
        "samtools view athal_wu_0_A.sort.bam \"Chr3:11,777,000-11,794,000\" | wc -l\n",
        "## 7081\n",
        "\n",
        "7. How many alignments show the read’s mate unmapped?\n",
        "samtools view athal_wu_0_A.sort.bam \"Chr3:11,777,000-11,794,000\" | cut -f7 | grep \"*\" | wc -l\n",
        "## 1983\n",
        "\n",
        "8. How many alignments contain a deletion (D)?\n",
        "samtools view athal_wu_0_A.sort.bam \"Chr3:11,777,000-11,794,000\" | cut -f6 | grep \"D\" | wc -l\n",
        "## 31\n",
        "\n",
        "9. How many alignments show the read’s mate mapped to the same chromosome?\n",
        "samtools view athal_wu_0_A.sort.bam \"Chr3:11,777,000-11,794,000\" | cut -f7 | grep \"=\" | wc -l\n",
        "## 4670\n",
        "\n",
        "10. How many alignments are spliced?\n",
        "samtools view athal_wu_0_A.sort.bam \"Chr3:11,777,000-11,794,000\" | cut -f6 | grep \"N\" | wc -l\n",
        "## 0\n",
        "\n",
        "11. How many sequences are in the genome file?\n",
        "12. What is the length of the first sequence in the genome file?\n",
        "13. What alignment tool was used?\n",
        "samtools view -H athal_wu_0_A.bam\n",
        "@HD VN:1.0 GO:none SO:coordinate\n",
        "@SQ SN:Chr1 LN:29923332 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:Chr2 LN:19386101 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:Chr3 LN:23042017 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:Chr4 LN:18307997 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:Chr5 LN:26567293 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:chloroplast LN:154478 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@SQ SN:mitochondria LN:366924 AS:wu_0.v7.fas SP:wu_0.v7.fas\n",
        "@RG ID:H100223_GAII05_0002 PL:SLX LB:wu_PII PI:400 DS:wu_0_Genome SM:wu_0\n",
        "@RG ID:Wii_PER01 PL:SLX LB:wu_phaseI PI:400 DS:wu_0_Genome SM:wu_0\n",
        "@RG ID:Wii_PER02 PL:SLX LB:wu_phaseI PI:400 DS:wu_0_Genome SM:wu_0\n",
        "@RG ID:Wii_SR03 PL:SLX LB:wu_phaseI PI:400 DS:wu_0_Genome SM:wu_0\n",
        "@PG ID:stampy\n",
        "\n",
        "## 7\n",
        "## 29923332\n",
        "## stampy\n",
        "\n",
        "14. What is the read identifier (name) for the first alignment?\n",
        "samtools view athal_wu_0_A.bam | head\n",
        "GAII05_0002:1:113:7822:3886#0 1187 Chr3 11699950 60 51M = 11700332 433 AAAAAAAATGTAAAACTGCTAAATCTCTCCTCTCTAAAGAACTCGTCCCCG CCCCCCBBBCCCCCCCCCCCCCCCCCCCCCCCCCCCBAAB??@ACBBCCCD PQ:i:21 SM:i:37 UQ:i:0 MQ:i:37 XQ:i:0 RG:Z:H100223_GAII05_0002\n",
        "\n",
        "## GAII05_0002:1:113:7822:3886#0\n",
        "\n",
        "15. What is the start position of this read’s mate on the genome? Give this as ‘chrom:pos’ if the read was mapped, or ‘*” if unmapped.\n",
        "## chr3:11700332\n",
        "\n",
        "16. How many overlaps (each overlap is reported on one line) are reported?\n",
        "bedtools bamtobed -i athal_wu_0_A.bam > athal_wu_0_A.bed\n",
        "bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | wc -l\n",
        "## 3101\n",
        "\n",
        "17. How many of these are 10 bases or longer?\n",
        "bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f16 > count.txt\n",
        "awk '$1>9{c++} END{print c+0}' count.txt\n",
        "## 2899\n",
        "\n",
        "18. How many alignments overlap the annotations?\n",
        "## 3101\n",
        "\n",
        "19. Conversely, how many exons have reads mapped to them?\n",
        "bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f4 | sort -u | wc -l\n",
        "## 21\n",
        "\n",
        "20. If you were to convert the transcript annotations in the file “athal_wu_0_A_annot.gtf” into BED format, how many BED records would be generated?\n",
        "bedtools intersect -wo -a athal_wu_0_A_annot.gtf -b athal_wu_0_A.bed | cut -f9 | cut -d \" \" -f4 | sort -u | wc -l\n",
        "## 4"
      ],
      "metadata": {
        "id": "A6X2GMhlhs3_"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fb6TY1QIhlQ-"
      },
      "outputs": [],
      "source": [
        ""
      ]
    }
  ]
}